### DISCUSSION TITLE
Assessing Feasibility for Improved Web Performance

### DISCUSSION THREAD

**Discussion by ramonmalcolm10** [inertiajs/inertia#1661] [Ideas]
@reinink would love to have your feedback on the following:

1. **HTML Streaming, React Server Components, and Prefetching**: Your perspective on how these techniques align with Inertia's architecture and whether there are any notable challenges in their implementation.

2. **Efficient Data Handling**: Would it be possible to only fetch data that are used in persistent layout once and apply lazy load technique on subsequent navigation.

3. **View Page Transition API**: Considering Inertia's structure, I'd love to know your thoughts on integrating the View Page Transition API for seamless transitions between views. How complex would this be to achieve?

**Comment by ramonmalcolm10** [comment]
Was just having a discussion on how the JS frameworks have evolved and shift to a model to support progressive enhancement. Would love to hear the community feedback on Inertia adapting this model as well.

**Comment by andresgutgon** [comment]
Any changes related with RSCs in inertia?

**Comment by aaersya-ardh** [comment]
Interesting discussion.

From a performance perspective, combining techniques like HTML streaming, server components, and smart prefetching can definitely reduce the perceived latency of page transitions.

One thing I’ve noticed in similar setups is that the real bottleneck sometimes shifts from network requests to rendering and layout updates in the browser, especially when components trigger frequent reflows or heavy paint operations.

For example, when testing page transitions and streaming responses, it's helpful to analyze:

• layout recalculations  
• paint and compositing layers  
• GPU acceleration behavior  
• frame timing in DevTools

I came across a breakdown that explains how browser rendering stages can influence perceived performance in modern web apps:
[grabslot](https://existo.tech/)

It might be useful when thinking about how these techniques interact with the browser's rendering pipeline.

**Comment by aaersya-ardh** [comment]
This is an interesting point about progressive enhancement and modern framework architecture.

In setups where server components and streaming are used, the initial response can feel faster, but sometimes the client-side rendering pipeline still becomes the limiting factor.

In some projects we observed performance issues not from the data fetching layer, but from:

- layout invalidation caused by component updates  
- expensive paint operations during transitions  
- compositing layers changing frequently

Profiling with the Performance tab in DevTools can reveal where frame drops actually occur.

There is also a good explanation of how rendering stages (layout → paint → compositing) affect performance in complex web apps:
[dewa86](https://www.brothertreehouse.com/)

It helped clarify why some apps still feel slow even after optimizing the network layer.
